Added Model Reference Adaptive Control in spinal attitude controller

makit0sh opened the pull request:
Added Model Reference Adaptive Control in spinal attitude controller.
This control can only be activated with service call on "/hydrus/spinal_mrac_trigger".

tongtybj commented on aerial_robot_nerve/spinal/mcu_project/Jsk_Lib/flight_control/attitude/attitude_control.cpp:
```diff
@@ -407,11 +442,203 @@ void AttitudeController::reset(void)
+
+void AttitudeController::yawFromPCallback(const spinal::YawFromPC& msg)
+{
+  pc_psi_ = msg.pc_yaw;
```
Ok, so you need the desired yaw angle and the current yaw angle.
The former one can be merged to `FourAxisCommand.msg`.
The later one should be sent independently.
Please at least change the message name `yawFromPC` to only `yaw` or `estimatedYaw`, not `FromPC`.

tongtybj commented on aerial_robot_control/src/control/pose_linear_controller.cpp:
```diff
@@ -328,12 +329,16 @@ namespace aerial_robot_control
   {
     /* ros publish */
     pid_pub_.publish(pid_msg_);
+    yaw_from_pc_pub_.publish(yaw_from_pc_msg_);
```
I recommend to create a new control class for MRAC called `AdaptiveController`, which is inherited from `PoseLinearController`, and merge your trigger python script into this new controller class

tongtybj reviewed (commented):
I add several comments to your codes. Please consider them, and discuss with me.

tongtybj commented:
@makit0sh 

Please add instruction to try your code in simulation.
For example, `roslaunch hydrus bringup.launch ...`.
Add rostest script like https://github.com/JSKAerialRobot/aerial_robot/blob/master/robots/hydrus/test/hydrus_control.test
is better

tongtybj commented:
Anyway, very clear commit. good job!

makit0sh commented:
I have refactored code based on advice of @tongtybj.

- created new plugin base adaptive controller for hydrus and merged python script
- renamed yaw topic

makit0sh commented:
You can use my MRAC only by specifying plugin name in FlightControl.yaml of robot config.

`aerial_robot_control_name: aerial_robot_control/hydrus_tilted_lqi_adaptive_controller`

tongtybj commented:
@makit0sh 

Please provide a bringup sample (e.g., in simulation):
For example, `roslaunch hydrus bringup.launch ....`

makit0sh commented:
@tongtybj 
I added new robot based on default_mode_201907.
The simulation launches with
`aerial_robot_control_name: aerial_robot_control/hydrus_tilted_lqi_controller`

tongtybj commented:
@makit0sh 

LGTM.

I will merge this PR after the experiment
